Major refactor

lann opened the pull request:
This change ended up changing a large percentage of the codebase. Some highlights:

- Reimplemented the lexer using the `logos` crate
- Resolved ambiguity between variant labels and keywords. Builtin keywords must now be '%'-prefixed in top-level value position.
- Resolved ambiguity between empty set and empty record values; empty records (with all fields set to `none`) are now represented as `{:}`.
- Added untyped parsing support, including the `UntypedValue` type and the `Node` AST representation.
- Added a "UI" test suite to ease test coverage expansion and migrated some integration tests to this new format.
- Replaced WasmValue::Error generic type with concrete WasmValueError.
- Removed completion support, which was more complex than it was worth

Fixes #18
Fixes #17
Fixes #6

lann committed:
Tweak error message formatting

lann committed:
Major refactor

This change ended up changing a large percentage of the codebase. Some
highlights:

- Reimplemented the lexer using the `logos` crate
- Resolved ambiguity between variant labels and keywords. Builtin
  keywords must now be '%'-prefixed in top-level value position.
- Resolved ambiguity between empty set and empty record values; empty
  records (with all fields set to `none`) are now represented as `{:}`.
- Added untyped parsing support, including the `UntypedValue` type and
  the `Node` AST representation.
- Added a "UI" test suite to ease test coverage expansion and migrated
  some integration tests to this new format.
- Replaced WasmValue::Error generic type with concrete WasmValueError.
- Removed completion support, which was more complex than it was worth

lann committed:
Update EBNF

lann committed:
Tweak error handling

sunfishcode commented on src/lex.rs:
```diff
@@ -1,185 +1,121 @@
+/// Represents a WAVE token.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Logos)]
+#[logos(error = LexingError)]
+#[logos(skip r"[ \t\r\n\f]+")]
```
I don't have an opinion either way, but including "\f" here differs from the current [Wit definition of whitespace](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md#whitespace) and it'd be nice for those to agree.

lann replied:
Done.

sunfishcode commented on tests/ui/accept-strings.wave:
```diff
@@ -0,0 +1,3 @@
+string("")
+string("WAVE🌊")
+string("\\\'\"\t\n\r")
```
This test doesn't currently include a string containing a Unicode escape.

lann replied:
Done.

sunfishcode commented on tests/ui/ui.wit:
```diff
@@ -0,0 +1,59 @@
+        ok,
+        err,
+        inf,
+        nan,
```
For completeness, it'd be good to have some non-keyword enum values here as well, so that the non-prefixed syntax can also be tested.

lann replied:
Done.

sunfishcode commented on src/lex.rs:
```diff
@@ -1,185 +1,121 @@
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Logos)]
+#[logos(error = LexingError)]
+#[logos(skip r"[ \t\r\n\f]+")]
+#[logos(subpattern label_word = r"[a-z][a-z0-9]*|[A-Z][A-Z0-9]*")]
```
Why does the label_word pattern include upper-case letters? It's my understanding that Wit identifiers are only lower-case alphanumeric.

lann replied:
Label words can be either all lowercase or all uppercase. I believe the idea is to allow retaining uppercase identifiers in bindings for languages that conventionally do that like Python.

https://github.com/WebAssembly/component-model/blob/main/design/mvp/Explainer.md
```ebnf
word          ::= [a-z] [0-9a-z]*
                | [A-Z] [0-9A-Z]*
```

fibonacci1729 replied:
FWIW, this is what WAC implements now as well (added here https://github.com/peterhuene/wac/pull/29).

sunfishcode replied:
Ah, I had forgotten about that. Looks good!

sunfishcode commented on src/parser.rs:
```diff
@@ -1,726 +1,451 @@
-        Ok(name.strip_prefix('%').unwrap_or(name))
+pub(crate) fn flattenable(kind: WasmTypeKind) -> bool {
+    use WasmTypeKind::*;
+    !matches!(kind, Variant | Enum | Option | Result)
```
With some/none/ok/err being keywords, is the plan still to make variant/enum flattenable? Or further, to make option-in-result or result-in-option flattenable? If so it'd be fine to do a TODO comment here for now.

lann replied:
Updated to allow inner variant/enum. I haven't thought through mixed option/result flattening yet so added a note as well.

sunfishcode reviewed (commented):
Wow, so much here! Yay for logos, yay for the AST, yay for untyped values, yay for resolving ambiguities, and yay for the new test framework :smile: .

lann committed:
Make enum and variant values flattenable

Now that option and result cases are reserved keywords there is no
longer any (potential) ambiguity with variant and enum cases.

lann committed:
Minor refactoring

lann committed:
Make whitespace consistent with WIT

lann committed:
Add tests/ui/README.md

...and change test input extension to `.waves`.

lann committed:
Update README

lann committed:
Make Node::as_char pub

lann committed:
Reorganize modules

...and some documentation improvements.

lann committed:
Upgrade wasmtime to 16

lann committed:
Remove LexingError

lann committed:
Make lex::Keyword pub

lann committed:
Re-export / alias required logos types

lann committed:
Rename repo to wasm-wave

To match the crate name.

lann committed:
Add UntypedFuncCall

lann committed:
Use semicolons to delimit UI test cases

lann committed:
Add non-keyword enum test cases

lann committed:
Add note about considering mixed option/result flattening

lann commented:
Other update highlights:

- Renamed repo to `wasm-wave` to match the crate
- Added `UntypedFuncCall` which holds AST for `<func-name>(<params...>)`
- Added trailing `;` to UI test cases
- Lots of doc updates
- Reorganized modules
